### ISSUE TITLE
Lua relation members gets post-way transform member tags

### ISSUE THREAD

**Issue by pnorman** [openstreetmap/osm2pgsql#659]
I am working on a transform which creates a table with wood areas, a name column, and no other transform-generated columns.

This means I want a way function like

```lua
function wood_ways(tags, num_keys)
    if tags["natural"] == wood then
        local cols = {}
        cols.name = tags["name"]
        return 0, cols, 1, 0
    end
    return 1, {}, 0, 0
end
```

This works properly, turning a way with `natural=wood` into a row with no transform-generated columns

If I am only concerned with old-style MPs, I can then add 

```lua
function wood_rels (tags, num_keys)
    if (tags["type"] == "multipolygon" and tags["natural"] == wood) then
        return 0, tags
    end
    return 1, {}
end

--- test function
-- @param tags OSM tags
-- @param member_tags OSM tags of relation members
-- @param member_roles OSM roles of relation members
-- @param membercount number of members
-- @return filter, cols, member_superseded, boundary, polygon, roads
function wood_rel_members (tags, member_tags, member_roles, membercount)
    assert(membercount == 2) -- hard-coded to avoid loops to generate member_tags
    if tags["natural"] == wood then
        local cols = {}
        cols.name = tags["name"]
        return 0, {}, {0, 0}, 0, 1, 0
    end
    return 1, {}, {0, 0}, 0, 0, 0
end
```

It's not obvious, but member_tags is member tags **after** transformation by the way function, i.e. columns to be output to PostgreSQL, not original tags.

This causes a problem when I ~~want~~need to handle old-style multipolygons, like a relation with `type=multipolygon` and two members, one with `natural=wood` and the other with no tags.

The output of `wood_ways` for the first way is `0, {}` and the second is `1, {}`. This means the function called for rel members is
```lua
wood_rel_members({type="multipolygon"}, {{}, {}}, {"",""}, 2)
```

This is the same as any two-member MP which does not have any tags on ways, which does not have natural=wood on any ways, or that does have natural=wood on the ways, but also has different tags (e.g. `leaf_type` key). Because of this, it's impossible to correctly handle old-style MPs with the above way transform.

Although I made this example with the multi backend, it shows up if you do the same thing with the pgsql backend. The difference is people tend to either **modify** existing keys when generating columns (e.g. @SomeoneElseOSM) or **add** new columns (e.g. @systemed), but seldom replace all the columns with the pgsql backend like I am doing.

I'm not certain if this is a bug in the Lua API we define, or what we've implemented.

Putting print statements into the way call shows that `wood_ways` gets once for each way, and once more for each relation membership a way participates in.

**Comment by pnorman** [comment]
My inclination is that this is a bug. I can probably work around it in ClearTables by encoding all the tags into one column as text, since I have a function to do that

```lua
--- Turns a lua table into a PostgreSQL hstore
-- @param table table to convert
-- @return table in hstore string format
function hstore (table)
    if table ~= nil then
        local s = ''
        for k,v in pairs(table) do
            s = s..',"'..string.gsub(string.gsub(k,'\\','\\\\'),'"', '\\"')..'"=>"'..string.gsub(string.gsub(v,'\\','\\\\'),'"', '\\"')..'"'
        end
        return string.sub(s, 2) -- clean up first ,
    end
end

function wood_ways(tags, num_keys)
    if tags["natural"] == wood then
        local cols = {}
        cols.name = tags["name"]
        cols.original_tags = hstore(tags)
        return 0, cols, 1, 0
    end
    return 1, {original_tags = hstore(tags)}, 0, 0
end
```

I'll then have the call ``wood_rel_members({type="multipolygon"}, {{original_tags='"natural"=>"wood"'}, {}}, {"",""}, 2)``, which contains the information required. But this feels like an ugly hack.

**Comment by pnorman** [comment]
> My inclination is that this is a bug. I can probably work around it in ClearTables by encoding all the tags into one column as text, since I have a function to do that

Nope, this isn't really an option. I was hoping that I'd be able to use avoid parsing that text and just using if it's different, but I need the tags in a table to check for each way if I want to accept the feature. The former just tells me if it's an old-style MP, not if I want it in the relevant table.

**Comment by pnorman** [comment]
Some relevant commits

https://github.com/openstreetmap/osm2pgsql/commit/7d9d8397
https://github.com/openstreetmap/osm2pgsql/commit/38d3e0cd

I think it's a bug, caused by it not exactly being clear what the pgsql backend is doing, which is what the multi backend is based on.

**Comment by lonvia** [comment]
There is a comment in the code [here](https://github.com/openstreetmap/osm2pgsql/blob/master/output-pgsql.cpp#L423-L427) confirming that the filtering is intentional. It was introduced in 2565bf738fe69b18443971c40fe4a142163a3714 and the reasoning there is consistent with the old C implementation.

The easy way to work around your problem is to add a fixed tag to filtered ways and check for that, i.e. like that:

```
function wood_ways(tags, num_keys)
    if tags["natural"] == wood then
        return 0, { name = tags["name"] } , 1, 0
    end
    return 1, { notwood = "" }, 0, 0
end

function wood_rel_members (tags, member_tags, member_roles, membercount)
  if new-style MP then
    -- do stuff
  else
    local iswood = false
    for i = 1,membercount do
      if member_tags["notwood"] == nil then
        iswood = true
        break
      end
    end
    if iswood then
       -- do more stuff
    end
end
```
Sidenote: this hack exploits the fact that we always unpack four return arguments for the function, even when the element is filtered. It might be a good idea for efficiency to reconsider that and not return tags when the element is marked as filtered.

While I don't think that the current behaviour is a bug (it serves the limited use-case of old-style MP filtering as done in the C transform well enough), I would agree that it is far from ideal. I'd be tempted to move the filtering into the LUA functions/C transform, so that it is left to the user to decide if the second filter step is even necessary. However, I don't know how many people rely on the current behaviour.

**Comment by pnorman** [comment]
> However, I don't know how many people rely on the current behaviour

I'm pretty sure no one does.

> The easy way to work around your problem is to add a fixed tag to filtered ways and check for that, i.e. like that:

The problem is that will lead to different interpretations on what is and isn't a MP depending on which table I'm looking at, so some object might be treated as an old-style MP for one table and an invalid MP for another. I want understandable and strict processing of old-style MPs so a tag change doesn't have an unexpected impact.

At least, that's what I think would happen, but it's hard to be sure.

> There is a comment in the code here confirming that the filtering is intentional. It was introduced in 2565bf7 and the reasoning there is consistent with the old C implementation.

Ya, but I'm not sure we ever thought about the implications of C and pgsql lua transforms seldom losing tagging information because they commonly only add tags, vs multi transforms where losing information is common with multiple tables.

**Comment by lonvia** [comment]
>>   However, I don't know how many people rely on the current behaviour

> I'm pretty sure no one does.

Then I would be in favour of changing pgsql output as well. It would allow to simplify the code (save copying around the tag list at least once).

**Comment by pnorman** [comment]
> Then I would be in favour of changing pgsql output as well. It would allow to simplify the code (save copying around the tag list at least once).

I'm not as sure about the pgsql output, I'll have to think about what's done there a bit more.

**Comment by pnorman** [comment]
Fixed in #721
